documentation/smg/16-serialiser.md

```diff
@@ -0,0 +1,275 @@
+0    |    tag     |      |      |     | Real Data | Imag Data
+1    | 32 + tag   | nElem|      |     | Real Data | Imag Data
+2    | 64 + tag   | dim1 | dim2 |     | Real Data | Imag Data
+N    | N<<5 + tag | dim1 | dim2 | ... | Real Data | Imag Data
+
+#### Sparse data
+Sparse numeric data (tags 29-31) are serialised as:
+
+RANK |   TYPE  |     tag  | dim1 | dim2 | Data | ... |    ...    | ...
+-----|---------|----------|------|------|:----:|:---:|-----------|------
+2    | Real    | 93       | dim1 | dim2 |   i  |  j  | Data      |
+2    | Bool    | 94       | dim1 | dim2 |   i  |  j  | Data      |
+2    | Complex | 95       | dim1 | dim2 |   i  |  j  | Real Data | Imag Data
+
+Based on `sparse(i,j,data)` [MATLAB documentation][sparse]
+
+__NB.__ only rank-2 sparse arrays are permitted in MATLAB
+
+__NB.__ Due to storing dimensions as `uint32` data, rather than `double`, sizes of sparse arrays are limited to (2^32)-1
+
+#### Struct data
+Structured tree data (tag 24) uses the standard header format:
+
+RANK |     tag    | dim1 | dim2 | ... | Data
+-----|------------|------|------|:---:|------
+NULL | 32 + tag   | 0 
+0    |    tag     |      |      |     | Data
+1    | 32 + tag   | nElem|      |     | Data
+2    | 64 + tag   | dim1 | dim2 |     | Data
+N    | N<<5 + tag | dim1 | dim2 | ... | Data
+
+
+Where data is formatted as follows:
+
+nFields | length(fieldName1) | length(fieldName2) | ... | fieldName1 | fieldName2 | ... | struct2cell(data)
+--------|--------------------|--------------------|:---:|------------|------------|:---:|------------------
+
+nFields and fieldName lengths are serialised as `uint32`
+
+__NB.__ For struct arrays `struct2cell` produces a Rank-(N+1) cell array, this means that for struct arrays, the limit on rank is 6.
+
+#### Cell array
+Cell array data (tag 23) are serialised as:
+
+RANK |     tag    | dim1 | dim2 | ... | Data
+-----|------------|------|------|:---:|------
+NULL | 32 + tag   | 0 
+0    |    tag     |      |      |     | Data
+1    | 32 + tag   | nElem|      |     | Data
+2    | 64 + tag   | dim1 | dim2 |     | Data
+N    | N<<5 + tag | dim1 | dim2 | ... | Data
+
+Where `data` is the concatenation of the serialisation of each element.
+
+#### Object Array
+Object array data (tag 23) are serialised as:
+
+RANK |     tag    | dim1 | dim2 | ... |Data|     ...            |    ...     |   ...   |  ...
+-----|------------|------|------|:---:|----|--------------------|------------|---------|------
+NULL | 32 + tag   | 0    |      |     | 33 | length(class_name) | class_name |         |
+0    |    tag     |      |      |     | 33 | length(class_name) | class_name | ser_tag | Data |
+1    | 32 + tag   | nElem|      |     | 33 | length(class_name) | class_name | ser_tag | Data |
+2    | 64 + tag   | dim1 | dim2 |     | 33 | length(class_name) | class_name | ser_tag | Data |
+N    | N<<5 + tag | dim1 | dim2 | ... | 33 | length(class_name) | class_name | ser_tag | Data |
+
+Where
```

Comment:
the first data col seems a bit redundant - you already have tag 32+23=55 for object array. It could be good to understand why this is useful.

Reply:
Mostly because it is serialised by a call to `serialise_simple_data`, which adds the tag and dims.

Reply:
Assuming that "serialise_simple_data" is your code rather than matlab code, as the code is not necessarily in front of the reader of this document as they read, a quick explanation of this would stop the reader puzzling over why the magic number 33 is here. Equally, as 33 is part of the data, it could just be omitted as it is not required to understand they header.